050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,806 @@
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 100Gi
+      class: regular-storage
+```
+
+#### Protecting `KafkaNodePool` resources from being shared between multiple Kafka clusters
+
+The `strimzi.io/cluster` label should guarantee that the node pool will match only one Kafka cluster.
+Additional protection will be made against undesired moves between different Kafka clusters when the `strimzi.io/cluster` label is modified.)
+The `.status` section of the `KafkaNodePool` resource will contain the Kafka cluster ID of the cluster it belongs to.
+For example:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: my-pool
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # ...
+status:
+  # ...
+  clusterId: cABpUKeDTqi7q_XOtBMgUv
+```
+
+The cluster ID is already part of the `Kafka` CR `.status` section.
+The operator will check the cluster IDs in every reconciliation and if the cluster IDs don't match between the `Kafka` and `KafkaNodePool` CRs, it will expect that there is some kind of an misconfiguration.
+It will throw an `InvalidresourceException` and wait for the user to fix the issue.
+
+### Scaling the node pools
+
+The `KafkaNodePool` resource will also support the `scale` sub-resource.
+Support for the `scale` sub-resource should allow integration with Kubernetes Horizontal Pod Autoscaler.
+That on its own doesn't really provide fully functional Kafka auto-scaling mechanism.
+For example, if the load is coming from a single partition which is handling too many messages, adding more nodes will not help.
+So a proper auto-scaling mechanism would need to understand the configuration or topics and partitions and their load to properly scale the broker.
+But the `scale` sub-resource would make it easier to develop such solutions in the future.
+
+### KRaft support
+
+To support the Kafka KRaft mode, a new field `role` will be added to the `KafkaNodePool` as well.
+It will contain a list of roles (`controller` or `broker`) which the nodes in given pool should have.
+This will allow to use different configurations.
+For example, for a small cluster with shared broker and controller responsibilities:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  namespace: my-namespace
+spec:
+  kafka:
+    version: 3.4.0
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.3"
+  # ...
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: mixed
+  namespace: my-namespace
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  role:
```

Comment:
maybe using the plural `rules` as it is in the Kafka configuration?

Reply:
Good point. We seem to use the plural form for list fields in our APIs as well. Thanks.